v2/internal/parse_test.go

```diff
@@ -28,23 +33,180 @@ func TestParseMessageFileBytes(t *testing.T) {
+		},
+		{
+			name: "nested test",
+			file: `{"nested": {"hello": "world"}}`,
+			path: "en.json",
+			messageFile: &MessageFile{
+				Path:   "en.json",
+				Tag:    language.English,
+				Format: "json",
+				Messages: []*Message{{
+					ID:    "nested.hello",
+					Other: "world",
+				}},
+			},
+		},
+		{
+			name: "basic test with description",
+			file: `{"notnested": {"description": "world"}}`,
+			path: "en.json",
+			messageFile: &MessageFile{
+				Path:   "en.json",
+				Tag:    language.English,
+				Format: "json",
+				Messages: []*Message{{
+					ID:          "notnested",
+					Description: "world",
+				}},
+			},
+		},
+		{
+			name: "basic test with id",
+			file: `{"key": {"id": "forced.id"}}`,
+			path: "en.json",
+			messageFile: &MessageFile{
+				Path:   "en.json",
+				Tag:    language.English,
+				Format: "json",
+				Messages: []*Message{{
+					ID: "forced.id",
+				}},
+			},
+		},
+		{
+			name: "basic test with description and dummy",
+			file: `{"notnested": {"description": "world", "dummy": "nothing"}}`,
+			path: "en.json",
+			messageFile: &MessageFile{
+				Path:   "en.json",
+				Tag:    language.English,
+				Format: "json",
+				Messages: []*Message{{
+					ID:          "notnested",
+					Description: "world",
+				}},
+			},
+		},
+		{
+			name: "deeply nested test",
+			file: `{"outer": {"nested": {"inner": "value"}}}`,
+			path: "en.json",
+			messageFile: &MessageFile{
+				Path:   "en.json",
+				Tag:    language.English,
+				Format: "json",
+				Messages: []*Message{{
+					ID:    "outer.nested.inner",
+					Other: "value",
+				}},
+			},
+		},
+		{
+			name: "multiple nested test",
+			file: `{"nested": {"hello": "world", "bye": "all"}}`,
+			path: "en.json",
+			messageFile: &MessageFile{
+				Path:   "en.json",
+				Tag:    language.English,
+				Format: "json",
+				Messages: []*Message{{
+					ID:    "nested.hello",
+					Other: "world",
+				}, {
+					ID:    "nested.bye",
+					Other: "all",
+				}},
+			},
+		},
+		{
+			name: "YAML nested test",
+			file: `
+outer:
+    nested:
+        inner: "value"`,
+			path:           "en.yaml",
+			unmarshalFuncs: map[string]UnmarshalFunc{"yaml": yaml.Unmarshal},
+			messageFile: &MessageFile{
+				Path:   "en.yaml",
+				Tag:    language.English,
+				Format: "yaml",
+				Messages: []*Message{{
+					ID:    "outer.nested.inner",
+					Other: "value",
+				}},
+			},
+		},
 	}
 	for _, testCase := range testCases {
 		actual, err := ParseMessageFileBytes([]byte(testCase.file), testCase.path, testCase.unmarshalFuncs)
-		if err != testCase.err {
-			t.Fatalf("expected error %#v; got %#v", testCase.err, err)
+		if (err == nil && testCase.err != nil) ||
+			(err != nil && testCase.err == nil) ||
+			(err != nil && testCase.err != nil && err.Error() != testCase.err.Error()) {
+			t.Errorf("%s failed: expected error %#v; got %#v", testCase.name, testCase.err, err)
+			continue
+		}
+		if actual == nil {
```

Comment:
I don't think this condition serves any purpose and therefore should be deleted.

Reply:
Actually yes, it is useful: when the `ParseMessageFileBytes()` call raises an error, `actual` is nil and the remaining lines of the test panic.

Would you prefer that I write `if err != nil` for the test instead? It is strictly equivalent, maybe more explicit. Another option would be to write the previous test as follows:
```go
if err != nil || testCase.err != nil {
	if err == nil || testCase.err == nil || err.Error() != testCase.err.Error()) {
		t.Errorf("%s failed: expected error %#v; got %#v", testCase.name, testCase.err, err)
	}
	continue
}
```
which forces to `continue` the for loop when an error is found, and makes the `actual` test useless.

What do you like best?

Reply:
Ah, I see what you mean. I think I would find it clearer written like this:

```go
if testCase.messageFile == nil && actual != nil || testCase.messageFile != nil && actual == nil {
    t.Fatalf("expected message file %#v; got %#v", testCase.messageFile, actual)
}

if testCase.messageFile != nil {
    if actual.Path != testCase.messageFile.Path {
        t.Fatalf("expected path %q; got %q", testCase.messageFile.Path, actual.Path)
    }
    if actual.Tag != testCase.messageFile.Tag {
        t.Fatalf("expected tag %q; got %q", testCase.messageFile.Tag, actual.Tag)
    }
    if actual.Format != testCase.messageFile.Format {
        t.Fatalf("expected format %q; got %q", testCase.messageFile.Format, actual.Format)
    }
    if !equalMessages(actual.Messages, testCase.messageFile.Messages) {
        t.Fatalf("expected %#v; got %#v", testCase.messageFile.Messages, actual.Messages)
    }
}
```

Reply:
Done, commit beea385